packages/zowe-explorer-api/src/profiles/UserSettings.ts

```diff
@@ -9,6 +9,16 @@
  *
  */
 
+import { ICreateDataSetOptions } from "@zowe/cli";
+
+export interface CreateDSOptionsWithTemplate extends ICreateDataSetOptions {
+    /**
+     * Name of data creation template
+     */
+    name?: string;
+}
+export type DataSetAllocTemplate = CreateDSOptionsWithTemplate;
```

Comment:
instead of adding a name property... would it make sense to rework the structure a little bit?

**Current**
```jsonc
    // ...
    "templates": [
      {
        "alcunit": "CYL",
        "blksize": 6160,
        "dirblk": 10,
        "dsorg": "PO",
        "primary": 1,
        "recfm": "FB",
        "lrecl": 80,
        "name": "test-dir-block-10"
      }
    ],
    // ...
```

**Proposed**
```jsonc
    // ...
    "templates": {
      "test-dir-block-10": {
        "alcunit": "CYL",
        "blksize": 6160,
        "dirblk": 10,
        "dsorg": "PO",
        "primary": 1,
        "recfm": "FB",
        "lrecl": 80
      }
    },
    // ...
```

Reply:
The other option would be to leverage the `zowe.config.json` which should make the templates available for someone using both CLI and ZE.

One approach is to create a separate profile type for templates which we could leverage in tandem with other profiles (like how `tso` and `zosmf` profiles are related)
```jsonc
               // ...
                "test-dir-block-10": {
                    "type": "ds-template",
                    "properties": {
                        "alcunit": "CYL",
                        "blksize": 6160,
                        "dirblk": 10,
                        "dsorg": "PO",
                        "primary": 1,
                        "recfm": "FB",
                        "lrecl": 80
                    },
                }
                // ... 
                // IMPORTANT: we should avoid default "ds-templates"
```

Another less appealing approach would be to associate the new template to a single `node/profile` 
```jsonc
        // ...
        "demo": {
            "type": "zosmf",
            "properties": {
                "host": "my.host.net",
                "port": 443,
            },
            "profiles": {
                "test-dir-block-10": {
                    "properties": {
                        "alcunit": "CYL",
                        "blksize": 6160,
                        "dirblk": 10,
                        "dsorg": "PO",
                        "primary": 1,
                        "recfm": "FB",
                        "lrecl": 80
                    },
            }
        },
        // ...
```

Reply:
@zFernand0 I would like to look at the first suggestion before merging this PR as it could change some things in the APIs.  For the second suggestion of config support of templates, I think it's a cool idea and wonder if should be handled at imperative or cli stage with wanting this to work for both cli and us.  Either way this one can be it's own followup story.

Reply:
> instead of adding a name property... would it make sense to rework the structure a little bit?
> 
> **Current**
> 
> ```js
>     // ...
>     "templates": [
>       {
>         "alcunit": "CYL",
>         "blksize": 6160,
>         "dirblk": 10,
>         "dsorg": "PO",
>         "primary": 1,
>         "recfm": "FB",
>         "lrecl": 80,
>         "name": "test-dir-block-10"
>       }
>     ],
>     // ...
> ```
> 
> **Proposed**
> 
> ```js
>     // ...
>     "templates": {
>       "test-dir-block-10": {
>         "alcunit": "CYL",
>         "blksize": 6160,
>         "dirblk": 10,
>         "dsorg": "PO",
>         "primary": 1,
>         "recfm": "FB",
>         "lrecl": 80
>       }
>     },
>     // ...
> ```

Pushed updates per this request, if tested before remove any templates in settings.json as the format of the object changed.
![Screen Shot 2023-05-24 at 2 37 32 PM](https://github.com/zowe/vscode-extension-for-zowe/assets/49491949/d1ce2771-4417-4319-8459-3d99b3b36821)

Reply:
@zFernand0 all the other history items are in arrays which is what I was trying to keep inline with

```
"zowe.ds.history": {
    "persistence": true,
    "favorites": [],
    "history": [],
    "templates": [],
    "sessions": [
      "zosmf"
    ],
    "searchHistory": []
  }
```

Reply:
Thanks for considering that... _other history items are in arrays_
That makes me wonder if templates should be inside the `zowe.ds.history` property or if they should have their own `zowe.ds.templates`.

I don't mind continuing discussions which is why I'm leaving a table to help us compare each individual approach.

That being said, I don't want to hold up the PR any longer.
I'm more than happy with the work and effort that was put into this already 😋

<table>
<thead>
<tr>
<td> 1️⃣ <strong>Initial - Flat Array</strong></td><td> 2️⃣ <strong>Current Structure</strong></td><td> 3️⃣ <strong>Proposed - Flat Object</strong></td>
</tr>
</thead>
<tbody>
<tr>
<td>

 ```jsonc
    // ...
    "templates": [
      {
        "alcunit": "CYL",
        "blksize": 6160,
        "dirblk": 10,
        "dsorg": "PO",
        "primary": 1,
        "recfm": "FB",
        "lrecl": 80,
        "name": "dirblk-10"
      },
      // ... more templates 
    ],
    // ...
    
```
</td>
<td>

```jsonc
    // ...
    "templates": [
      {
        "dirblk-10": {
          "alcunit": "CYL",
          "blksize": 6160,
          "dirblk": 10,
          "dsorg": "PO",
          "primary": 1,
          "recfm": "FB",
          "lrecl": 80
        }
      },
      // ... more templates
    }
    // ...
```
</td>
<td>

```jsonc
    // ...
    "templates": {
      "dirblk-10": {
        "alcunit": "CYL",
        "blksize": 6160,
        "dirblk": 10,
        "dsorg": "PO",
        "primary": 1,
        "recfm": "FB",
        "lrecl": 80
      },
      // ... more templates
    },
    // ...
    
    
```
</td>
</tr>
<tr>
<td>

```typescript
this.mDsTemplates.filter((element) => {
  // templateName = element.name 
  // ... element.dirblk ...
})
```
</td>
<td>

```typescript
this.mDsTemplates.filter((template) => {
  Object.entries(template).forEach(([key, value]) => {
    // templateName = key;
  });
  // ... continue operations ...
  // ... template[templateName].dirblk ...
  // Note: we may be required to check for 
  // `template.hasOwn(templateName)`
})
```
</td>
<td>

```typescript
for (const [name, template] of Object.entries(this.mDsTemplates)) {
  // templateName = name;
  // ... template.dirblk ...
})
```
</td>
</tr>
</tbody>
</table>

Again, all different ways to reach the same goal. 😋